### ISSUE TITLE
abap2UI5 & ajson

### ISSUE THREAD

**Issue by oblomov-dev** [abap2UI5/abap2UI5#843]
Heads up!

abap2UI5 is now running per default with ajson instead of /ui2/cl_json, expect some bugs, please test your app and open an issue.

Reason for the change:
* ajson give the possibility to cleanup the whole json handling & payload creation of abap2UI5 now, complexity is a lot reduced now! (no more manual assigns etc.)
* structure are send per default as objects to the frontend
* attribute names have no restrictions in their length anymore
* no more redundant usage of rtti for payload handling
* no more nested call chains and loops (everything is done in ajson)
* no dependency to /ui2/json anymore (allows downport compatibility)

At the moment both logics are part of abap2UI5, you can switch back with the following importing parameter (but I hope I can delete this option as soon as possible):
https://github.com/abap2UI5/abap2UI5/blob/c9e154d02e4cdf6395ea88a7eb70a486cdd162eb/src/02/z2ui5_cl_fw_http_handler.clas.abap#L10

Or use the latest tag:
https://github.com/abap2UI5/abap2UI5/releases/tag/1.117.0

If you run into problems open an issue or check this section (all logic is now done here):
(1) frontend to backend
https://github.com/abap2UI5/abap2UI5/blob/c9e154d02e4cdf6395ea88a7eb70a486cdd162eb/src/01/02/z2ui5_cl_fw_model_ajson.clas.abap#L155

(2) backend to frontend
https://github.com/abap2UI5/abap2UI5/blob/c9e154d02e4cdf6395ea88a7eb70a486cdd162eb/src/01/02/z2ui5_cl_fw_model_ajson.clas.abap#L66

Compression handling changed, in the future there is hopefully just the standard compression needed and additionally custom compression can be defined, as you can see in this example:
https://github.com/abap2UI5/abap2UI5-samples/blob/main/src/z2ui5_cl_demo_app_148.clas.abap

The corresponding custom compression is implemented here:
https://github.com/abap2UI5/abap2UI5/blob/c9e154d02e4cdf6395ea88a7eb70a486cdd162eb/src/01/05/z2ui5_cl_cc_chartjs.clas.abap#L616

Please check the documentation, a2json is a very nice framework and easy to use:
https://github.com/sbcgua/ajson

Best regards.

**Comment by choper725** [comment]
what doea the custom comression parameter do? i see in chartjs it points to the class

**Comment by oblomov-dev** [comment]
you can import the name of your class which implements your compression method. in the example i put it into the cusotm control of char.js because this specific custom controls needs this compression mehtod. i hope it is easier to maintain like this and we are not dependent from the /ui2/_json behaviour anymore.

**Comment by choper725** [comment]
still looking, trying to find where is that custom compression for chartjs...

found this in the binding class
![image](https://github.com/abap2UI5/abap2UI5/assets/129428114/cdb4bc85-3ae8-4b02-a70d-b246393bb3eb)

does this 30 len check is relevant with ajson ??

**Comment by oblomov-dev** [comment]
Thank you for checking this!

The 30 length check is not needed anymore, it is not active with the ajson logic.

yes its very deep in the strucutres:
<img width="679" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/78ba9954-da62-439c-8d90-b37f661eba7d">

maybe it was before wrong and is now correct?

we can also now adjust the structure or make something snake_case to make it fit.

besides that:
all ypur normal apps and other stuff you did with abap2UI5 works so far? (asking because its a big change, i checked all demos, but i want to make sure)

**Comment by choper725** [comment]
yes i checked the things i did and so far works as expected..

i think it worked good before, as i got what i expected in the json of every chart,

i think the logic of compressing based of values is wrong as it is possible to set those values intentionaly..

but i think the issue now that its not showing its because somet values are not getting hidden in the current logic,

the scale->x/y/r->grid->borderdash .

**Comment by oblomov-dev** [comment]
nice this sounds good!

then there is only the chart.js issue left but i think it is fixable. i adjusted again the logic slightly and hope it is now easier to use and understand why i think the new approach is better and more flexible for upcoming custom controls.

you can import now with the binding two instances:
(1) custom_mapping -> you can influence the output of the fronend json
(2) custom_filter -> like old compress function
you can use the following out-of-the-box options:
(1)
<img width="348" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/af412a53-099d-48d3-a294-04ad2641f17b">
(2)
<img width="264" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/c1613534-ad84-4cf2-9ccc-99e1559de095">

Or of course develop your own ones, means implement the interface and import the instance with the binding method as it is done with the chart.js custom control in demo 148.

I adjusted the examples, check out:
z2ui5_cl_demo_app_090
z2ui5_cl_demo_app_148
z2ui5_cl_demo_app_153

Next step for me is then to delete the old logic and all /ui2/cl_json dependencies.

Best regards.

**Comment by choper725** [comment]
looks nice, need to research it more.

regarding demo chartjs (z2ui5_cl_demo_app_148)

it seems no compress happen at all after recent fix

**Comment by oblomov-dev** [comment]
ah forgot to import the custom filter, adjusted it now with https://github.com/abap2UI5/abap2UI5-samples/commit/ca0196fb3d4c94b732b4b09cab9ccc5dbd11b8d7

**Comment by choper725** [comment]
nope, still getting this values for some reason which are not a plugin nor defined in demo 148...
![image](https://github.com/abap2UI5/abap2UI5/assets/129428114/6372f276-9192-4bcb-808d-147bb5e453cb)


seems like not all values are getting hidden..

**Comment by oblomov-dev** [comment]
ah ok, fixed it now with the latest PR here:
https://github.com/abap2UI5/abap2UI5/blob/0c7130cfc241b86b8664ffc6c672a90c8821776d/src/01/05/z2ui5_cl_cc_chartjs.clas.abap#L636

Now i get the UI again:
<img width="800" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/f2259c15-decb-412e-b6dc-dcddc248acdb">

But still bugs in the frontend -> backend communication.

Additonally deleted the old /ui2/cl_json dependency and the old importing parameter compress_mode and pretty_mode.

**Comment by choper725** [comment]
great ill check, but i think this still need word as its possible to send false as a value, and in the current logic it will be hidden instead of being sent with false value

**Comment by oblomov-dev** [comment]
nice, ok but we are getting to it! i added now also the option to set a mapper for the frontend -> backend communication with:
https://github.com/abap2UI5/abap2UI5/pull/847

check out demo 153:
https://github.com/abap2UI5/abap2UI5-samples/blob/fbd326b1bda4292f1a8004c09585c3420a0b19ed/src/z2ui5_cl_demo_app_153.clas.abap#L80

Now the app or the custom control can influence the complete process of mapping (before: formatting) and filtering (before: compression). abap2UI5 just calls the imported methods and if nothing is set uses its default mapping and filtering. Besides that abap2UI5 has no own logic anymore which is a big reduction in complexity.

**Comment by oblomov-dev** [comment]
@choper725 fyi deleted the /ui2/cl_json dependency from driver.js and replaced with ajson:
https://github.com/abap2UI5/abap2UI5/blob/main/src/01/05/z2ui5_cl_cc_driver_js.clas.abap

**Comment by choper725** [comment]
great, i see you working on imagemap also, there is also the spreadsheet,

one question tho, how is the abap_bool handling in this.? as this was trouble before with ui2

**Comment by oblomov-dev** [comment]
yes imagemapster is now updated. 
https://github.com/abap2UI5/abap2UI5/pull/860

abap_bool handling works fine, it is mapped to true or false. and if you need specific behaviour you can esily implement your own mapper and import it with the _bind method.

**Comment by choper725** [comment]
great, ill looked at ajson seems great implementation of json handling,
ill play around with it with chartjs as it seems to be the mosr complex one

**Comment by choper725** [comment]
something is weird when updating...
the binded struc. are empty in abap....

**Comment by oblomov-dev** [comment]
updated the demo 148 again. i think the problem was that no mapper was set for the way back, i added now to_snake_case.
no field symbol error anymore, roundtrip works, but the output is not exactly what it was before but very close.

**Comment by choper725** [comment]
setting it back meaning getting it to how it was binded ?? seems alot flexible yet complex as before it was auto.

**Comment by oblomov-dev** [comment]
Yes when you do a specific mapping for the way back -> front you also have to set the way front -> back. Abap2UI5 acts now just as a middleware, before it was auto but with a lot of side effects. Now you have more manual work but you have complete freedom and control.

**Comment by oblomov-dev** [comment]
Took the chance and made a lot of refactoring: https://github.com/abap2UI5/abap2UI5/pull/862

The name of the http handler changed to:
z2ui5_cl_http_handler
And the importing parameter is now a single structure. don't forget to adjust your handler method.

As always, let me know if you find bugs.

**Comment by choper725** [comment]
> Took the chance and made a lot of refactoring: #862
> 
> The name of the http handler changed to: z2ui5_cl_http_handler And the importing parameter is now a single structure. don't forget to adjust your handler method.
> 
> As always, let me know if you find bugs.

thanks for updating, those are some breaking changes

**Comment by choper725** [comment]
> updated the demo 148 again. i think the problem was that no mapper was set for the way back, i added now to_snake_case. no field symbol error anymore, roundtrip works, but the output is not exactly what it was before but very close.

i think the issue is that the mapper back should go through the filter also...

**Comment by oblomov-dev** [comment]
ah i think there was another bug with the mapper, but anyway adding filer was a good idea. fixed and added with https://github.com/abap2UI5/abap2UI5/pull/865

seems to work now, just one chart on the second page, but i am not sure how it was before

**Comment by choper725** [comment]
fixed the error on my side and created a PR,

you are right something with the mapper is wrong, but i adjusted my code on the chartjs cc

what i found is that once i map a "snake" variable - i.e data_x_y_r.
its not working correctly once calling from front2back.

then i notich that variables with one snake, work correcly,
so i changed data_x_y_r to data_radial and now its working.

**Comment by oblomov-dev** [comment]
thank you, big help, good to see that it works again! chart.js was a lot of struggle with the old way, i think this approach is now flexible enough, especially when other custom controls and frontend libraries in the future are needed, i hope this way will work for a lot of use cases.

yes there might also be issues with the mapping, feel free to open issues if you run into problems.

<img width="800" alt="image" src="https://github.com/abap2UI5/abap2UI5/assets/102328295/7c66d634-0bef-49de-b3bb-d4eee2de92d1">

**Comment by oblomov-dev** [comment]
this issue got big, i close it, just open a new issue if you find more bugs.
